### PR TITLE
tools: create `diagnose_tensorboard`

### PR DESCRIPTION
**Pull request by wchargin**
Summary:
Users often report problems that depend on environment-specific
configuration. Rather than asking them to find all this information and
enter it into an issue template manually, we can ask them to run a
script and paste its output verbatim into the issue. Furthermore, we can
detect and suggest fixes to common problems, such as #1907 and #2010.
The script can grow as we see fit to add new diagnoses and suggestions.

Test Plan:
The script is designed to be robust to errors in each individual
diagnosis: in the worst case, it prints a stack trace and continues to
the next one. I’ve manually tested that the main framework works in
Python 2 and 3 on Linux and in Python 3 on Windows.

Automated testing of this script is possible, but would take a fair
number of CPU cycles to run tests (setting up virtualenvs and Conda,
installing and importing TensorFlow many times). Given that this script
is never a production dependency and is explicitly designed to be run in
a discussion context, light testing seems reasonable.

To simulate a bad hostname, add

```python
socket.getfqdn = lambda: b"\xc9".decode("utf-8")
```

to the top of the script.

To simulate a bad `.tensorboard-info` directory and test the quoting
behavior, run

```shell
export TMPDIR="$(mktemp -d)/uh oh/" &&
mkdir -p "${TMPDIR}/.tensorboard-info" &&
chmod 000 "${TMPDIR}/.tensorboard-info" &&
python ./tensorboard/tools/diagnose_tensorboard.py
```

To cross-check the autoidentification logic:

```
$ python tensorboard/tools/diagnose_tensorboard.py |
> awk '/version / { print $NF }'
e093841ffaea564cb2410e0b430bd0c552ada208
$ git hash-object tensorboard/tools/diagnose_tensorboard.py
e093841ffaea564cb2410e0b430bd0c552ada208
$ git rev-parse HEAD:tensorboard/tools/diagnose_tensorboard.py
e093841ffaea564cb2410e0b430bd0c552ada208
$ git cat-file blob e093841ffaea564cb2410e0b430bd0c552ada208 |
> diff -u tensorboard/tools/diagnose_tensorboard.py - | wc -l
0
```

wchargin-branch: diagnose-me

### CODE REVIEW

**Comment by manivaradarajan** [comment]
Yay, thanks for starting this.

File: tensorboard/tools/diagnose_me.py [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+#     again". Should be imperative with no trailing punctuation. May
+#     contain inline Markdown.
+#   description (str): A full enumeration of the steps that the user
+#     should take to accept the suggestion. Within this string, prose
+#     should be formatted with `reflow`. May contain Markdown.
+Suggestion = collections.namedtuple("Suggestion", ("headline", "description"))
+
+
+def check(fn):
+  """Register a function as a check.
```
Comment (manivaradarajan) [inline tensorboard/tools/diagnose_me.py]: Can you add a line that this is intended to be used as a decorator, i.e., as @check.

File: tensorboard/tools/diagnose_me.py [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+  set_up_logging()
+
+  print("### Diagnostics")
+  print()
+
+  markdown_code_fence = "``````"  # seems likely to be sufficient
+  print(markdown_code_fence)
+  suggestions = []
+  for check in CHECKS:
+    print("--- check: %s" % check.__name__)
```
Comment (stephanwlee) [inline tensorboard/tools/diagnose_me.py]: Consider putting one newline after to give some space between the checks (alternatively, we can use the color but that complicates things I guess).

Comment (wchargin) [reply tensorboard/tools/diagnose_me.py]: Okay, if you prefer. (I like the concise form, but don’t feel strongly.)
I did consider color, but decided against it—color is easy on xterm
emulators, but this script needs to work on Windows, too.

File: tensorboard/tools/diagnose_me.py (hunk 1 above)
Comment (wchargin) [reply tensorboard/tools/diagnose_me.py]: Sure.

**Pull request merged**